Sources/AMQPClient/ChannelHandlers/AMQPConnectionMultiplexHandler.swift

```diff
@@ -11,282 +11,366 @@
+                        properties: properties,
+                        redelivered: deliver.redelivered,
+                        body: body
+                    ),
+                    for: deliver.consumerTag
+                )
+            case let .return(`return`):
+                channel.eventHandler?.receiveReturn(.init(
+                    replyCode: `return`.replyCode,
+                    replyText: `return`.replyText,
+                    exchange: `return`.exchange,
+                    routingKey: `return`.routingKey,
+                    properties: properties,
+                    body: body
+                ))
+            default:
+                // TODO: take down channel
+                preconditionUnexpectedFrame(frame)
             }
         case .heartbeat:
             let heartbeat = Frame(channelID: frame.channelID, payload: .heartbeat)
-            context.writeAndFlush(self.wrapOutboundOut(.frame(heartbeat)), promise: nil)
+            context.writeAndFlush(wrapOutboundOut(.frame(heartbeat)), promise: nil)
         }
     }
 
-    func openChannel(id: Frame.ChannelID) -> EventLoopFuture<AMQPChannelHandler<AMQPConnectionMultiplexHandler>> {
-        if let channel = self.channels[id] {
-            return self.eventLoop.makeSucceededFuture(channel)
-        }
-
-        return self.write(outband: .frame(.init(channelID: id, payload: .method(.channel(.open(reserved1: ""))))))
-            .flatMapThrowing  { response in 
-                guard case .channel(let channel) = response, case .opened(let channelID) = channel, channelID == id else {
-                    throw AMQPConnectionError.invalidResponse(response)
-                }
-
-                let channelHandler = AMQPChannelHandler(parent: self, channelID: channelID, eventLoop: self.eventLoop)
-
-                self.channels[channelID] = channelHandler
-
-                return channelHandler
-            }
-    }
-
-    func close(reason: String = "", code: UInt16 = 200) -> EventLoopFuture<Void> {
-        return self.write(outband: .frame(.init(
-            channelID: 0, payload: .method(.connection(.close(.init(replyCode: code, replyText: reason, failingClassID: 0, failingMethodID: 0)))))))
-            .flatMapThrowing { response in
-                guard case .connection(let connection) = response, case .closed = connection else {
-                    throw AMQPConnectionError.invalidResponse(response)
+    func write(context: ChannelHandlerContext, data: NIOAny, promise: EventLoopPromise<Void>?) {
+        let (outbound, responsePromise) = unwrapOutboundIn(data)
+        switch state {
+        case let .error(e), let .blocked(e):
+            promise?.fail(e)
+            responsePromise?.fail(e)
+            return
+        case .unblocked:
+            if let responsePromise {
+                guard let channelId = outbound.channelId else { preconditionFailure("Response expected without channel id") }
```

Comment:
Is this message correct - why "without" ?

Reply:
well... the API here is a bit implicit (I did not want to touch too many things at once, but we can definitely refactor this in a future PR)
basically, you cannot send a raw frame (without a channel id) and expect a response. I'll try to reword it more nicely.

This could only happen with an "empty" `AMPQOutbound.bulk([])` or a `.bytes(...)` - so this would definitely be an implementation error.